air/src/execution_step/lambda_applier/applier.rs

```diff
@@ -132,31 +157,74 @@ fn select_by_path_from_canon_map<'value>(
+            select_by_path_from_canon_map_stream(canon_stream_iter, &body_part, exec_ctx)?
         }
         (Err(..), Some(canon_stream)) => {
-            let value = canon_stream.as_jvalue();
-            Cow::Owned(value)
+            // csm.$.key case
+            let prefix_with_path = false;
+            let tetraplet = update_tetraplet_with_path(canon_map.tetraplet(), original_lambda, prefix_with_path);
+            let value = Cow::Owned(canon_stream.as_jvalue());
+
+            MapLensResult::from_cow(value, tetraplet)
         }
         _ => {
-            let SecurityTetraplet {
-                peer_pk,
-                service_id,
-                function_name,
-                json_path,
-            } = canon_map.tetraplet().as_ref();
-            let json_path = json_path.to_string() + &prefix.to_string();
-
-            let tetraplet = SecurityTetraplet::new(peer_pk, service_id, function_name, json_path).into();
-            let value = CanonStream::new(vec![], tetraplet).as_jvalue();
-            Cow::Owned(value)
+            // csm.$.non_existing_key case
+            let prefix_with_path = false;
+            let tetraplet = update_tetraplet_with_path(canon_map.tetraplet(), original_lambda, prefix_with_path);
+            let value = CanonStream::new(vec![], tetraplet.clone()).as_jvalue();
+            let value = Cow::Owned(value);
+
+            MapLensResult::from_cow(value, tetraplet)
         }
     };
     Ok(result)
 }
 
+fn exctract_idx<'lambda>(
+    lambda: &'lambda NonEmpty<ValueAccessor<'_>>,
+    exec_ctx: &ExecutionCtx<'_>,
+) -> ExecutionResult<(usize, &'lambda [ValueAccessor<'lambda>])> {
+    let (prefix, body) = lambda.split_first();
+    let idx = match prefix {
+        ValueAccessor::ArrayAccess { idx } => *idx,
+        ValueAccessor::FieldAccessByName { field_name } => {
+            return lambda_to_execution_error!(Err(LambdaError::FieldAccessorAppliedToStream {
+                field_name: field_name.to_string(),
+            }));
+        }
+        ValueAccessor::FieldAccessByScalar { scalar_name } => {
+            let scalar = exec_ctx.scalars.get_value(scalar_name)?;
+            lambda_to_execution_error!(try_scalar_ref_as_idx(scalar))?
+        }
+        ValueAccessor::Error => unreachable!("should not execute if parsing succeeded. QED."),
+    };
+    Ok((idx as usize, body))
+}
+
+fn update_tetraplet_with_path(
+    original_tetraplet: &RcSecurityTetraplet,
+    original_path: &impl Display,
+    prefix_with_path: bool,
+) -> RcSecurityTetraplet {
+    let SecurityTetraplet {
+        peer_pk,
+        service_id,
+        function_name,
+        json_path,
+    } = original_tetraplet.as_ref();
+
+    let json_path = if prefix_with_path {
+        json_path.to_string() + &original_path.to_string()
+    } else {
+        original_path.to_string()
+    };
+
+    SecurityTetraplet::new(peer_pk, service_id, function_name, json_path).into()
```

Comment:
BTW, struct update syntax would make this function shorter.  But I am ambivalent to it.